lib/src/native/realm_core.dart

```diff
@@ -1509,80 +1495,64 @@ class _RealmCore {
-      return;
-    }
+  static void _onConnectionStateChange(Object userdata, int oldState, int newState) {
+    final controller = userdata as SessionConnectionStateController;
 
     controller.onConnectionStateChange(ConnectionState.values[oldState], ConnectionState.values[newState]);
   }
 
   Future<void> sessionWaitForUpload(Session session) {
     final completer = Completer<void>();
-    _realmLib.realm_dart_sync_session_wait_for_upload_completion(
-      session.handle._pointer,
-      Pointer.fromFunction(_sessionWaitCompletionCallback),
-      completer.toPersistentHandle(),
-      _realmLib.addresses.realm_dart_delete_persistent_handle,
-      scheduler.handle._pointer,
-    );
+    final callback = Pointer.fromFunction<Void Function(Handle, Pointer<realm_sync_error_code_t>)>(_sessionWaitCompletionCallback);
+    final userdata = _realmLib.realm_dart_userdata_async_new(completer, callback.cast(), scheduler.handle._pointer);
+    _realmLib.realm_sync_session_wait_for_upload_completion(session.handle._pointer, _realmLib.addresses.realm_dart_sync_wait_for_completion_callback,
+        userdata.cast(), _realmLib.addresses.realm_dart_userdata_async_free);
     return completer.future;
   }
 
   Future<void> sessionWaitForDownload(Session session) {
     final completer = Completer<void>();
-    _realmLib.realm_dart_sync_session_wait_for_download_completion(
-      session.handle._pointer,
-      Pointer.fromFunction(_sessionWaitCompletionCallback),
-      completer.toPersistentHandle(),
-      _realmLib.addresses.realm_dart_delete_persistent_handle,
-      scheduler.handle._pointer,
-    );
+    final callback = Pointer.fromFunction<Void Function(Handle, Pointer<realm_sync_error_code_t>)>(_sessionWaitCompletionCallback);
+    final userdata = _realmLib.realm_dart_userdata_async_new(completer, callback.cast(), scheduler.handle._pointer);
+    _realmLib.realm_sync_session_wait_for_download_completion(session.handle._pointer, _realmLib.addresses.realm_dart_sync_wait_for_completion_callback,
+        userdata.cast(), _realmLib.addresses.realm_dart_userdata_async_free);
     return completer.future;
   }
 
-  static void _sessionWaitCompletionCallback(Pointer<Void> userdata, Pointer<realm_sync_error_code_t> errorCode) {
-    try {
-      final completer = userdata.toObject<Completer<void>>(isPersistent: true);
-      if (completer == null) {
-        return;
-      }
+  static void _sessionWaitCompletionCallback(Object userdata, Pointer<realm_sync_error_code_t> errorCode) {
+    final completer = userdata as Completer<void>;
 
-      if (errorCode != nullptr) {
-        // Throw RealmException instead of RealmError to be recoverable by the user.
-        completer.completeError(RealmException(errorCode.toSyncError().toString()));
-      } else {
-        completer.complete();
-      }
-    } finally {
-      _realmLib.realm_free(errorCode.cast());
+    if (errorCode != nullptr) {
+      completer.completeError(errorCode.toSyncError());
```

Comment:
This will throw SyncError instead of RealmException. Why are we doing this drastic change here? There is a comment that we really want to throw RealmException at this point.

Reply:
That was a bad merge on my part, I'll fix it.

Reply:
Reverted. Though now that I look at it I'm not sure why we're wrapping the `SyncError` message in a `RealmException` and ignoring the error code and category members - wouldn't those be useful to users handling sync errors?